File: invoice-agent-java/src/main/java/com/agent/liquidalts/invoiceagent/extraction/PdfTextExtractor.java
package com.agent.liquidalts.invoiceagent.extraction;

import org.apache.pdfbox.Loader;
import org.apache.pdfbox.pdmodel.PDDocument;
import org.apache.pdfbox.text.PDFTextStripper;
import org.springframework.stereotype.Component;

import java.io.IOException;
import java.io.UncheckedIOException;

/**
 * Deterministic text extraction — the LLM never sees raw bytes, only text.
 * For scanned invoices, plug an OCR engine (Tesseract/Textract) in behind
 * this same interface.
 */
@Component
public class PdfTextExtractor {

    public String extractText(byte[] pdfBytes) {
        try (PDDocument document = Loader.loadPDF(pdfBytes)) {
            PDFTextStripper stripper = new PDFTextStripper();
            stripper.setSortByPosition(true);
            return stripper.getText(document);
        } catch (IOException e) {
            throw new UncheckedIOException("Failed to extract text from PDF", e);
        }
    }
}

File: invoice-agent/src/main/kotlin/com/agent/liquidalts/invoiceagent/extraction/PdfTextExtractor.kt
package com.agent.liquidalts.invoiceagent.extraction

import org.apache.pdfbox.Loader
import org.apache.pdfbox.text.PDFTextStripper
import org.springframework.stereotype.Component

/**
 * Deterministic text extraction — the LLM never sees raw bytes, only text.
 * For scanned invoices, plug an OCR engine (e.g. Tesseract/Textract) in
 * behind this same interface.
 */
@Component
class PdfTextExtractor {

    fun extractText(pdfBytes: ByteArray): String =
        Loader.loadPDF(pdfBytes).use { document ->
            PDFTextStripper().apply { sortByPosition = true }.getText(document)
        }
}
